src/hooks/usePortal.ts

```diff
@@ -0,0 +1,303 @@
+  approveInputToken: UseMutateAsyncFunction;
+  shouldApprove: unknown;
+  gaslessSignature: `0x${string}` | undefined;
+  inputToken: string;
+  inputTokenDecimals: number;
+  inputAmount: string;
+  outputToken: string;
+  sender: string;
+};
+
+export type PortalQuoteResponse = {
+  tx: TransactionToSubmit;
+  context: PortalQuoteContext;
+};
+
+export type PortalQuoteContext = {
+  orderId: string;
+  minOutputAmount: string;
+  minOutputAmountUsd: number;
+  slippageTolerancePercentage: number;
+  gasLimit: string;
+  inputAmount: string;
+  inputAmountUsd: number;
+  inputToken: string;
+  outputToken: string;
+  outputAmount: string;
+  outputAmountUsd: number;
+  partner: string;
+  feeToken: string;
+  feeAmount: string;
+  feeAmountUsd: number;
+  sender: `0x${string}`;
+  recipient: `0x${string}`;
+  target: `0x${string}`;
+  value: string;
+  route: string[];
+  steps: string[];
+};
+
+export type PortalTransactionResult =
+  | { hash: string; signature?: undefined }
+  | { signature: string; hash?: undefined };
+
+const PORTAL_API_BASE_URI = "https://api.portals.fi/v2";
+const PORTALS_CHAIN_ID = {
+  [mainnet.id]: "ethereum",
+  [arbitrum.id]: "arbitrum",
+  [optimism.id]: "optimism",
+  [fantom.id]: "fantom",
+};
+
+const useCheckApproval = (
+  sender: `0x${string}`,
+  inputToken: `0x${string}`,
+  inputAmount: string,
+  inputTokenDecimals: number,
+) => {
+  const chain = useChain();
+
+  return useQuery({
+    queryKey: [
+      "portalCheckApproval",
+      sender,
+      inputToken,
+      inputAmount,
+      inputTokenDecimals,
+    ],
+    queryFn: async () => {
+      const url = new URL(`${PORTAL_API_BASE_URI}/approval`);
+      url.searchParams.append("sender", sender);
+      url.searchParams.append(
+        "inputToken",
+        `${PORTALS_CHAIN_ID[chain.id]}:${inputToken}`,
+      );
+      url.searchParams.append(
+        "inputAmount",
+        `${parseUnits(inputAmount || `0`, inputTokenDecimals)}`,
+      );
+
+      const response = await fetch(url.toString(), { method: "GET" });
+
+      if (!response.ok) {
+        throw new Error(
+          `Error executing call to portal checkApproval: ${response.status} ${response.statusText}`,
+        );
+      }
+
+      const portalRouterApprovalData = await response.json();
+
+      return {
+        shouldApprove: portalRouterApprovalData.context.shouldApprove,
+        canPermit: portalRouterApprovalData.context.canPermit,
+        routerAddress: portalRouterApprovalData.context.target,
+        approveTx: portalRouterApprovalData?.approve,
+        permit: portalRouterApprovalData?.permit,
+      };
+    },
+  });
+};
+
+const useApproveInputToken = ({
+  approveTx,
+  permit,
+  canPermit,
+  shouldApprove,
+}: {
+  approveTx: TransactionToSubmit;
+  permit: PermitTx;
+  canPermit: boolean;
+  shouldApprove: boolean;
+}) => {
+  const { sendTransactionAsync } = useSendTransaction();
+  const { data: walletClient } = useWalletClient();
+
+  return useMutation({
```

Comment:
Im thinking we can rely on wagmi here.
Hooks to be used: 
- `usePrepareTransactionRequest`;
- `useSendTransaction` - you can pass prepared request there;
- `useSignTypedData`.

This is good because we check transaction success first.

This is similar design as we do with all other transactions where we first simulateWriteContract and then writeContract